### ISSUE TITLE
Confusing error message when a `void` value is used

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#56992]
```dart
FutureOr<void> foo() => print(""); // print() returns void
```
The code above produces an error in both analyzer and VM. Analyzer error message is `A value of type 'void' can't be returned from the function 'foo' because it has a return type of 'FutureOr<void>'` and VM error is `Error: A value of type 'void' can't be returned from a function with return type 'FutureOr<void>'.`

This error message is confusing. `void` is assignable to `FutureOr<void>`, the problem here is that a `void` value returned by `print()` is used. Please update the error message accordingly.

cc @eernstg 

`Dart SDK version: 3.7.0-71.0.dev (dev) (Fri Oct 25 17:06:43 2024 -0700) on "windows_x64"`

**Comment by eernstg** [comment]
Right. I adjusted the labels to indicate that this is concerned with the diagnostics from both tools. I removed `type-bug` because i think `improve-diagnostics` is usually specified alone, and it is taken to imply some kind of a `type-bug` treatment.

The error message sounds like there was a subtype test `void <: FutureOr<void>`, and it failed (so we can't return that value because it isn't type correct). However, as @sgrekhov mentions, that subtype relation _does_ hold.

The actual reason why there is an error is that the allowlist for using the result from evaluation of an expression of type `void` (https://github.com/dart-lang/language/blob/9dc3737010f3ccac5ef54bf63b402d8e86b9115c/specification/dartLangSpec.tex#L22666) doesn't contain any items that match `return print("");`.

So I'd expect an error message saying something like "This expression has a type of 'void' so its value can't be used" or "This expression has type 'void' and can't be used".

**Comment by Wdestroier** [comment]
Can the implementation or specification be fixed? It should work. I asked Gemini 2.5 Pro and it agrees:

```
User:
Will this code work in Dart?

`FutureOr<void> foo() => print(""); // print() returns void`

Agent:
Yes, that code will work perfectly fine in Dart.

...

Since print("") returns void, and void is one of the valid types for a function declared to return FutureOr<void>, the code is type-safe and will compile and run without issues.
In summary: Your function foo() synchronously executes print("") and implicitly returns void, which satisfies the FutureOr<void> return type.
```

**Comment by bwilkerson** [comment]
> So I'd expect an error message saying something like "This expression has a type of 'void' so its value can't be used" or "This expression has type 'void' and can't be used".

Would we want the same message when using an explicit `return` (as below), or is the current message better in that situation?
```dart
import 'dart:async';

FutureOr<void> foo() {
  return print("");
}
```

If the current message would be better for the case above, then perhaps we should simply enhance the current message in order to help build the correspondence in the user's mind. Perhaps something like "A value of type 'void' can't be implicitly returned from a function with return type 'FutureOr<void>'.".
